presto-native-execution/presto_cpp/main/expression/RowExpressionOptimizer.cpp

```diff
@@ -0,0 +1,709 @@
+          whenArgs.emplace_back(getConstantRowExpression(constantExpr));
+        }
+      } else {
+        VELOX_USER_CHECK(!exprInputs[i]->inputs().empty());
+        const auto& matchExpr = exprInputs[i]->inputs().back();
+        whenArgs.emplace_back(
+            veloxExprToRowExpression(matchExpr, inputWhenArgs[0]));
+      }
+
+      whenArgs.emplace_back(
+          veloxExprToRowExpression(caseValue, inputWhenArgs[1]));
+      result.emplace_back(getWhenSpecialForm(whenArgs, i));
+    }
+  }
+
+  // Else clause.
+  if (numInputs % 2 != 0) {
+    result.push_back(veloxExprToRowExpression(
+        exprInputs[numInputs - 1], inputArgs[numArgs - 1]));
+  }
+  return {result, true};
+}
+
+json RowExpressionConverter::getSpecialForm(
+    const exec::ExprPtr& expr,
+    const json& input) {
+  json res;
+  res["@type"] = "special";
+  std::string form;
+  if (input.contains("form")) {
+    form = input["form"];
+  } else {
+    // If input json is a call expression instead of a special form, for cases
+    // like 'is_null', the key 'form' will not be present in the input json.
+    form = expr->name();
+  }
+  // Presto requires the field form to be in upper case.
+  std::transform(form.begin(), form.end(), form.begin(), ::toupper);
+  res["form"] = form;
+  auto exprInputs = expr->inputs();
+  res["arguments"] = json::array();
+
+  // Arguments for switch expression include special form expression 'when'
+  // so it is constructed separately. If the switch expression evaluation found
+  // a case that always evaluates to true, the second value in pair switchResult
+  // will be false and the first value in pair will contain the value
+  // corresponding to the simplified case.
+  if (form == "SWITCH") {
+    auto switchResult = getSwitchSpecialFormArgs(expr, input);
+    if (switchResult.second) {
+      res["arguments"] = switchResult.first;
+    } else {
+      return switchResult.first.front();
+    }
+  } else {
+    json::array_t inputArguments = input["arguments"];
+    const auto numInputs = exprInputs.size();
+    VELOX_USER_CHECK_LE(numInputs, inputArguments.size());
+    for (auto i = 0; i < numInputs; i++) {
+      res["arguments"].push_back(
+          veloxExprToRowExpression(exprInputs[i], inputArguments[i]));
+    }
+  }
+  res["returnType"] = getTypeSignature(expr->type());
+
+  return res;
+}
+
+json RowExpressionConverter::toConstantRowExpression(
+    const exec::ExprPtr& expr) {
+  json res;
+  auto constantExpr = std::dynamic_pointer_cast<const exec::ConstantExpr>(expr);
+  VELOX_USER_CHECK_NOT_NULL(constantExpr);
+  auto constantRowExpr = getConstantRowExpression(constantExpr);
+  protocol::to_json(res, constantRowExpr);
+  return res;
+}
+
+json RowExpressionConverter::toCallRowExpression(
+    const exec::ExprPtr& expr,
+    const json& input) {
+  json res;
+  res["@type"] = "call";
+  protocol::Signature signature;
+  std::string exprName = expr->name();
+  if (veloxToPrestoOperatorMap_.find(expr->name()) !=
```

Comment:
exprName can be used in both places where expr->name() is used in this if block.